qemu/qmp.go

```diff
@@ -916,6 +916,18 @@ func (q *QMP) ExecuteVFIODeviceAdd(ctx context.Context, devID, bdf string) error
 	return q.executeCommand(ctx, "device_add", args, nil)
 }
 
+// ExecuteVFIOMediatedDeviceAdd adds a VFIO mediated device to a QEMU instance
+// using the device_add command. devID is the id of the device to add.
+// Must be valid QMP identifier. sysfsdev is the VFIO mediated device.
+func (q *QMP) ExecuteVFIOMediatedDeviceAdd(ctx context.Context, devID, sysfsdev string) error {
```

Comment:
I don't see the need for a new function here. Instead, you should modify the existing function that you previously introduced:
```go
func (q *QMP) ExecutePCIVFIOMediatedDeviceAdd(ctx context.Context, devID, sysfsdev, addr, bus string) error {
 	args := map[string]interface{}{
 		"id":       devID,
 		"driver":   "vfio-pci",
 		"sysfsdev": sysfsdev,
 	}
 	if bus != "" {
 		args["bus"] = bus
 	}
        if addr != "" {
                args["addr"] = addr
        }
 	return q.executeCommand(ctx, "device_add", args, nil)
}
```

Reply:
In kata-containers/runtime#594, qmp.ExecuteVFIODeviceAdd() is used to hotplug VFIO device on root bus, instead of PCI bridge, so I implement the corresponding qmp.ExecuteVFIOMediatedDeviceAdd() here.

Reply:
@xindazhao I haven't realized what was done in kata-containers/runtime#594, but I still think we should rework the existing function, rather than duplicating with a new one.
@rbradford @amshinde wdyt ?

Reply:
Yes, I agree, its better to rework existing function.

Reply:
hi @amshinde, do you mean using qmp.ExecutePCIVFIOMediatedDeviceAdd() when hotplug VFIO device on root bus(kata-runtime: virtcontainers/qemu.go), instead of qmp.ExecuteVFIODeviceAdd() ? 
How to calculate the addr value for VFIO device on root bus may be problem.

Reply:
We need to be on the lookout for hardcode -pci suffices due to #43

Reply:
@rbradford yes you're right. But keeping function name with `PCI` in it makes explicit the fact that we're going to use `vfio-pci` here, right?

@xindazhao is there anything unclear in my [comment](https://github.com/intel/govmm/pull/47/commits/97fb33ae398f8428eeff40f0e068d0d02146f8f7#r215008309)? Let me know, I'll clarify ;)

Reply:
Hi @sboeuf, it's possible to use function qmp.ExecutePCIVFIOMediatedDeviceAdd() to do hotplug on root bus, but the value of parameter addr is hard to determine, an empty slot on root bus is expected to be automatically picked, I tried several methods in kata-runtime but failed.

Using q.qmpMonitorCh.qmp.ExecutePCIVFIOMediatedDeviceAdd(q.qmpMonitorCh.ctx, devID, device.SysfsDev, "0xf", "pci.0");
Everything is ok.

Using q.qmpMonitorCh.qmp.ExecutePCIVFIOMediatedDeviceAdd(q.qmpMonitorCh.ctx, devID, device.SysfsDev, "0xf", "");
Everything is ok.

Using q.qmpMonitorCh.qmp.ExecutePCIVFIOMediatedDeviceAdd(q.qmpMonitorCh.ctx, devID, device.SysfsDev, "", "pci.0")
# docker run -it --runtime=kata-runtime --rm  --device=/dev/vfio/0 -v /dev:/dev  xinda/ubuntu /bin/bash   
docker: Error response from daemon: OCI runtime create failed: QMP command failed: unknown.

Using q.qmpMonitorCh.qmp.ExecutePCIVFIOMediatedDeviceAdd(q.qmpMonitorCh.ctx, devID, device.SysfsDev, "", "")
# docker run -it --runtime=kata-runtime --rm  --device=/dev/vfio/0 -v /dev:/dev  xinda/ubuntu /bin/bash   
docker: Error response from daemon: OCI runtime create failed: QMP command failed: unknown.

Reply:
@xindazhao I am confused here. The patch that I proposed should have exactly the same behavior as the one you proposed with this PR, hence I think you should have the same issue with your patch, right?
Now, that being said, I think Kata Containers device model holds the addresses already used, so from the code, you should be able to specify the next address available. cc @amshinde

Reply:
@xindazhao Agree with @sboeuf. The revised patch should have the same behaviour as the earlier one, as providing bus="" and addr="", should result in hotplugging on the root bus. Qemu would find an empty slot on the root bus.
@xindazhao Can you confirm if your previous patch works correctly?

Reply:
Sorry, it's my fault. I didn't look at the patch carefully, I missed the function below.
        if addr != "" {
                args["addr"] = addr
        }
I mistakenly thought that simply use functions ExecutePCIVFIOMediatedDeviceAdd() to replace functions ExecuteVFIOMediatedDeviceAdd.

I will update the PR ASAP.